test/cpp/end2end/xds/xds_pick_first_end2end_test.cc

```diff
@@ -0,0 +1,136 @@
+#include <grpcpp/support/status.h>
+
+#include "src/core/ext/filters/client_channel/backup_poller.h"
+#include "src/core/ext/filters/client_channel/lb_policy/xds/xds_channel_args.h"
+#include "src/core/ext/filters/client_channel/resolver/fake/fake_resolver.h"
+#include "src/core/lib/address_utils/sockaddr_utils.h"
+#include "src/core/lib/config/config_vars.h"
+#include "src/core/lib/gprpp/env.h"
+#include "src/proto/grpc/testing/xds/v3/cluster.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/pick_first.pb.h"
+#include "test/core/util/scoped_env_var.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/end2end/connection_attempt_injector.h"
+#include "test/cpp/end2end/xds/xds_end2end_test_lib.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+using ::envoy::extensions::load_balancing_policies::pick_first::v3::PickFirst;
+
+class PickFirstTest : public XdsEnd2endTest {
+ protected:
+  absl::variant<Status, size_t> WaitForAnyBackendHit(size_t start, size_t end) {
+    Status status;
+    size_t index = 3;
+    SendRpcsUntil(DEBUG_LOCATION, [&](const RpcResult& result) -> bool {
+      if (!result.status.ok()) {
+        status = result.status;
+        return false;
+      }
+      for (size_t i = start; i < end; ++i) {
+        if (backends_[i]->backend_service()->request_count() > 0) {
+          backends_[i]->backend_service()->ResetCounters();
+          index = i;
+          return false;
+        }
+      }
+      return true;
+    });
+    if (!status.ok()) {
+      return status;
+    }
+    return index;
+  }
+};
+
+// Run both with and without load reporting, just for test coverage.
+INSTANTIATE_TEST_SUITE_P(
+    XdsTest, PickFirstTest,
+    ::testing::Values(XdsTestType(), XdsTestType().set_enable_load_reporting()),
+    &XdsTestType::Name);
+
+TEST_P(PickFirstTest, PickFirstConfigurationIsPropagated) {
+  grpc_core::testing::ScopedExperimentalEnvVar env_var(
+      "GRPC_EXPERIMENTAL_PICKFIRST_LB_CONFIG");
+  CreateAndStartBackends(6);
+  // Change cluster to use pick_first with shuffle option.
+  auto cluster = default_cluster_;
+  PickFirst pick_first;
+  pick_first.set_shuffle_address_list(true);
+  cluster.mutable_load_balancing_policy()
+      ->add_policies()
+      ->mutable_typed_extension_config()
+      ->mutable_typed_config()
+      ->PackFrom(pick_first);
+  balancer_->ads_service()->SetCdsResource(cluster);
+  bool saw_different_backend = false;
+  size_t start_index = 0;
+  for (size_t i = 0; i < 100 && !saw_different_backend; ++i) {
+    // Update EDS resource.  This will send a new address list update to the LB
+    // policy.
+    balancer_->ads_service()->SetEdsResource(BuildEdsResource(
+        EdsResourceArgs({{"locality0", CreateEndpointsForBackends(
+                                           start_index, start_index + 3)}})));
+    auto result = WaitForAnyBackendHit(start_index, start_index + 3);
+    size_t backend_index = absl::get<size_t>(result);
```

Comment:
If the variant holds a status, then this will crash.  I think you need to check if it failed before extracting the value.

The code here will become much cleaner once you switch to using `StatusOr<>`:

```
auto backend_index = WaitForAnyBackendHit(start_index, start_index + 3);
ASSERT_TRUE(backend_index.ok()) << backend_index.status();
if (*backend_index != start_index) return;
```